Remove empty lines from view_definition to avoid unecessary recreation of resources

alvaroqueiroz opened the pull request:
## Changes
<!-- Summary of your changes that are easy to understand -->

This PR solves the problem described here: https://github.com/databricks/terraform-provider-databricks/pull/3330

empty new lines on view_definition cause terraform to plan to change the view definition.
This PR adds the removal of empty lines to `DiffSuppressFunc` to avoid this problem

## Tests
<!-- 
How is this tested? Please see the checklist below and also describe any other relevant tests 
-->

- [x] `make test` run locally
- [ ] relevant change in `docs/` folder
- [x] covered with integration tests in `internal/acceptance`
- [ ] relevant acceptance tests are passing
- [ ] using Go SDK

![image](https://github.com/databricks/terraform-provider-databricks/assets/23335136/5f4a6192-cdf3-43b5-93ae-1a81850445f5)

![image](https://github.com/databricks/terraform-provider-databricks/assets/23335136/f664acb6-5012-4b24-8b04-c0d07353bea9)

Commit:
remove empty new lines to test for state change

alvaroqueiroz commented:
hello @mgyucht do you think this is ready for review?

mgyucht reviewed (commented):
This looks good to me. @alexott Do you have any concerns with this?

Only thing is that we should have one integration test to verify this behaves as expected and prevent regression. Look at `internal/acceptance/sql_table_test.go`. If you install the Databricks CLI (https://github.com/databricks/cli) and login to your workspace, you should be able to run any test. Can you try to add a test for view definition with multiple lines?

alexott commented:
My main concern is things like this:

```
line1\nline2" -> "line1line2
```

This will break view definitions like this, where it's split between multiple lines:

```
select col1
from table
```

will become:

```
select col1from table
```

This really should be discussed with the SQL engineering team if it makes sense to do it.

But from the error description, the issue is caused by empty lines, not by newlines, so the replacement should be from this regex `\n{1,}` -> `\n`. Also it makes sense to check if it's empty line, vs line with spaces only.

alvaroqueiroz commented:
hello @alexott,
this change would not break line1\nline2" -> "line1line2
I have even a test for that

Commit:
add acceptance test

Commit:
improve SuppressDiffWhitespaceAndEmptyLines logic

alvaroqueiroz commented:
@mgyucht I have added one acceptance test.
I have added two steps, one creates the view and the second just adds spaces and empty lines to the view definition. It did not generate alter tables as it would with SuppressDiffWhitespaceChange

Let me know if I should make changes to this test

alvaroqueiroz commented:
hello @mgyucht 
do you think it is worth proceeding with the review of this PR or I should close it?
I just wanted to fix this bug, but if the team thinks this is not the best solution, I'm open to discussing one that fits

nkvuong commented:
@alvaroqueiroz could you resolve the merge conflict for this one?

For this change, could you add a unit test that validates there would not be a change when there is diff between `InstanceState` and `HCL`?

Commit:
add unit test

alvaroqueiroz commented:
done @nkvuong 
let me know if you need anything else

nkvuong commented:
@alvaroqueiroz the conflict resolution for `internal/acceptance/sql_table_test.go` needs a small fix - could you take a look?

alvaroqueiroz commented:
@nkvuong I have synced my fork with the master,
the conflict is still present?

alvaroqueiroz commented:
@nkvuong fixed

nkvuong commented:
this test `TestResourceSqlTableUpdateView_IgnoreNewlineInDefinition` is failing

Commit:
fix tests attempt

alvaroqueiroz commented:
@nkvuong I think I'm not setting up the test correctly for some reason.
It seems to me that the Update flag will trigger an update regardless of the supressdiff method.

Even If I use the old `SuppressDiffWhitespaceChange` with just space changes, it will alter the table.
Do you have tips on how I could fix my tests?

nkvuong commented:
@alvaroqueiroz for unit testing, you could follow the examples of `TestCatalogSuppressCaseSensitivity`, which tests that the Diff calculated by Terraform is correct

alvaroqueiroz commented:
just fyi @nkvuong the test is passing because of [this line ](https://github.com/databricks/terraform-provider-databricks/pull/3361/files#diff-72a91d237a7ff752570e5ff10f2245c6f66e0014ace5439104cdcb94f40cb9a4R536)

I will get the test with ExpectedDiff in place

alvaroqueiroz commented:
@nkvuong I have tested with ExpectedDiff, I see the same behaviour.
Not even the `SuppressDiffWhitespaceChange` (the one in prod today) works on the tests

![image](https://github.com/databricks/terraform-provider-databricks/assets/23335136/8d2ff900-46f9-4fee-bc52-8c66a40683a2)

![image](https://github.com/databricks/terraform-provider-databricks/assets/23335136/60df3dd5-bcb0-4eab-9266-e2818704f5c9)

```python
func TestResourceSqlTableUpdateView_IgnoreNewlineInDefinition(t *testing.T) {
	qa.ResourceFixture{
		InstanceState: map[string]string{
			"name":            "barview",
			"catalog_name":    "main",
			"schema_name":     "foo",
			"table_type":      "VIEW",
			"cluster_id":      "existingcluster",
			"view_definition": "SELECT * FROM main.foo.bar",
		},
		ExpectedDiff: map[string]*terraform.ResourceAttrDiff{
			"catalog_name": {
				Old:         "main",
				New:         "main",
				RequiresNew: false,
			},
			"cluster_id": {
				Old:         "existingcluster",
				New:         "existingcluster",
				RequiresNew: false,
			},
			"name": {
				Old:         "barview",
				New:         "barview",
				RequiresNew: false,
			},
			"schema_name": {
				Old:         "foo",
				New:         "foo",
				RequiresNew: false,
			},
			"table_type": {
				Old:         "VIEW",
				New:         "VIEW",
				RequiresNew: false,
			},
			"view_definition": {
				Old:         "SELECT * FROM main.foo.bar",
				New:         "SELECT * FROM main.foo.bar",
				NewComputed: false,
				RequiresNew: false,
				NewRemoved:  false,
				Sensitive:   false,
			},
			"column.#": {
				Old:         "",
				New:         "",
				NewComputed: true,
				RequiresNew: true,
			},
			"properties.%": {
				Old:         "",
				New:         "",
				NewComputed: true,
				RequiresNew: false,
			},
		},
		HCL: `
		name               = "barview"
		catalog_name       = "main"
		schema_name        = "foo"
		table_type         = "VIEW"
		cluster_id         = "existingcluster"
		view_definition    = "SELECT * FROM main.foo.bar "
		`,
		Resource: ResourceSqlTable(),
		ID:       "main.foo.barview",
	}.ApplyNoError(t)
}
```

Commit:
add test with ExpectedDiff

vsluc commented:
@alvaroqueiroz Appreciate your work. I see a lot of value in this fix, as this unnecessary detection of changes increases the `apply` time significantly. Also, the recreation of the views with no change is not desirable, as it could run into contention with active ETLs that consume data from the views.
